proto/ptypes/controller.proto

```diff
@@ -48,6 +48,7 @@ message Volume {
 
 message ReplicaAddress {
     string address = 1;
+    string instanceName = 2;
```

Comment:
Tracing through the code, I see that the purpose of adding this field is to help with the case:
1. longhorn-manager want the engine to do something with one of the replica (for example, adding)
2. longhorn-manager would send this info to engine  
3. engine would use this one to make sure it is talking to the correct replica instance (among its multiple replicas' instances)

This is working well if the operation just involves 1 replica. However, for an operation that involves multiple replicas such as PrepareRebuildReplica or ReplicaVerifyRebuild this is maybe not complete as:
1. The operation requires the engine to talk to multiple replicas: the target replica and the source replica
2. For the target replica, the engine knows the instance name from the parameter ReplicaAddress.InstanceName so it is good
3. However, for the source replica, engine doesn't know the instance name thus it skips the validation mechanism by setting empty string https://github.com/longhorn/longhorn-engine/pull/902/files#diff-d4c1f609d8fdb5a2cd78e82cd073ba59ec0ad3779ff8d86b935293e08b422646R56

In order to completely verify all replicas, I am thinking if we should make the engine to remember the instance name of all if its replicas. AKA adding instanceName [here](https://github.dev/longhorn/longhorn-engine/blob/d1d6d308d37a5b465bc1c2824da2e0702830de7d/pkg/types/types.go#L130-L133). WDYT?

Reply:
You are right in everything you have posted here. We often DO skip instance name validation because the engine does not know the instance name of its replicas, and it WOULD be better if it did. The v2 engine is being written to include replica instance name as a "first class" identifier that the engine keeps track of.

However, when I set out to complete this work, I made refactoring longhorn-engine in this way a non-goal (https://github.com/longhorn/longhorn/blob/master/enhancements/20230420-engine-identity-validation.md#non-goals) because of the added complexity. IMO, the work I have done so far will fix the all observed real-world issues. 

Because:

- Replicas of the same volume rarely exist within the same instance-manager (and thus, are not susceptible to the issue in which one might quickly take the port of another, and
- Communicating with a different replica of the same volume would not incorrectly expand one of the replicas, etc,

The problem-set we would be addressing by doing as you suggest would be small compared to the work required. IMO we should complete this work as-is and potentially add a future enhancement to take it further if there is support to do so.

Reply:
Thank you for the explanation. Agree that it is not worth the effort